### PR TITLE
fix(client): tx forwarding at epoch boundary

### PR DESCRIPTION
**Pull request by mikhailOK**
Before this change tx forwarding works the following way:
- If we are a chunk producer at `(height+1)..=(height+4)` and current
epoch, save it in our own pool and don't forward.
- Otherwise forward to the chunk producer for `(height+3)` and current epoch.

The problem is that if we decide not to forward and a new epoch starts,
the tx gets lost: #2216

This change adds a heuristic:
- If we are a chunk producer at `(height+1)..=(height+4)` and current
epoch, and the planned new epoch start is in `0..=(height+4)`,
we both save the tx and forward to a chunk producer for the next epoch.

Test plan
---------
sanity/transactions.py now passes
sanity/state_sync_routed.py seems to pass
sanity/rpc_query.py passes with test change
test_catchup_random_single_part_sync_skip_15 passes with test change

### CODE REVIEW

**Comment by SkidanovAlex** [comment]
Can you extend the heuristic to:

```
If we are *not* a chunk producer at (height+1)..=(height+4) and current
epoch, and the planned new epoch start is in 0..=(height+4),
we forward to a chunk producer for the next epoch.
```

?

File: chain/client/src/client.rs [hunk 1]
```diff
@@ -1085,21 +1082,58 @@ impl Client {
...
+    ) -> Result<NetworkClientResponses, Error> {
+        let head = self.chain.head()?;
         let me = self.validator_signer.as_ref().map(|vs| vs.validator_id());
         let shard_id = self.runtime_adapter.account_id_to_shard_id(&tx.transaction.signer_id);
-        let cur_block_header = unwrap_or_return!(
-            self.chain.get_block_header(&head.last_block_hash),
-            NetworkClientResponses::NoResponse
-        )
-        .clone();
+        let cur_block_header = self.chain.get_block_header(&head.last_block_hash)?.clone();
```
Comment (bowenwang1996) [inline chain/client/src/client.rs]:
```suggestion
        let cur_block_header = self.chain.get_head_header()?.clone();
```

File: chain/client/src/client.rs [hunk 2]
```diff
@@ -1110,24 +1144,22 @@ impl Client {
...
             || self.runtime_adapter.will_care_about_shard(me, &head.last_block_hash, shard_id, true)
         {
-            let gas_price = unwrap_or_return!(
-                self.chain.get_block_header(&head.last_block_hash),
-                NetworkClientResponses::NoResponse
-            )
-            .inner_rest
-            .gas_price;
+            let gas_price =
+                self.chain.get_block_header(&head.last_block_hash)?.inner_rest.gas_price;
```
Comment (bowenwang1996) [inline chain/client/src/client.rs]: Why do you get the header again?

File: chain/client/src/client.rs [hunk 3]
```diff
@@ -1085,21 +1082,58 @@ impl Client {
         );
 
         // Send message to network to actually forward transaction.
-        self.network_adapter.do_send(NetworkRequests::ForwardTx(validator, tx));
+        self.network_adapter.do_send(NetworkRequests::ForwardTx(validator, tx.clone()));
 
-        NetworkClientResponses::RequestRouted
+        Ok(())
```
Comment (bowenwang1996) [inline chain/client/src/client.rs]: Why not return `NetworkClientResponses` here?

File: pytest/lib/cluster.py [hunk 4]
```diff
@@ -408,6 +409,7 @@ def init_cluster(num_nodes, num_observers, num_shards, config, genesis_config_ch
                  for line in err.decode('utf8').split('\n') if '/test' in line]
     assert len(node_dirs) == num_nodes + num_observers, "node dirs: %s num_nodes: %s num_observers: %s" % (len(node_dirs), num_nodes, num_observers)
 
+    print("Search for stdout and stderr in %s" % node_dirs)
```
Comment (bowenwang1996) [inline pytest/lib/cluster.py]: remove?

Comment (mikhailOK) [reply pytest/lib/cluster.py]: why remove?

File: chain/client/src/client.rs (hunk 3 above)
Comment (mikhailOK) [reply chain/client/src/client.rs]: because we can only return NetworkClientResponses::RequestRouted

File: pytest/lib/cluster.py (hunk 4 above)
Comment (bowenwang1996) [reply pytest/lib/cluster.py]: Oh I thought it is some debug output. Nvm

File: chain/client/src/client.rs (hunk 3 above)
Comment (bowenwang1996) [reply chain/client/src/client.rs]: Why is that a reason for not returning it? It removes the need to return `NetworkClientResponses` from the caller.

Comment (SkidanovAlex) [reply chain/client/src/client.rs]: @bowenwang1996 I'm with Misha here. If the outcome of the function is always the same except for errors, it should be `Result<(), Error>`.

**Pull request merged**